### PR TITLE
Add `mergeStepFunctionAndLambdaTraces` flag to merge step function and lambda traces

### PR DESCRIPTION
**Pull request by kimi-p**
### What does this PR do?
Help customer JsonMerge their payload with Step Function Context Object when flag `mergeStepFunctionAndLambdaTraces` is set to `true`.

The logic is as follows:
- For the serverless framework generated JSON file, we loop over all resources and find step functions.
- For a step function definition, we loop over every steps in the state machine definition and find steps using default lambda api (as oppose to legacy lambda api).
- If the step do not have payload field defined, or the field is defined with value `"$"` (default value), we inject lambda payload with step function context.

### Motivation
Customers sometimes have tens of lambda functions in a state machine. This PR helps them merge their lambda traces into step functions trace by setting one custom flag in the `serverless.yaml`.

### Testing Guidelines

Tested in ddserverless org.
image

### Types of changes

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Misc (docs, refactoring, dependency upgrade, etc.)

### Check all that apply

- [ ] This PR's description is comprehensive
- [ ] This PR contains breaking changes that are documented in the description
- [ ] This PR introduces new APIs or parameters that are documented and unlikely to change in the foreseeable future
- [ ] This PR impacts documentation, and it has been updated (or a ticket has been logged)
- [ ] This PR's changes are covered by the automated tests
- [ ] This PR collects user input/sensitive content into Datadog

### CODE REVIEW

File: src/env.spec.ts [hunk 1]
```diff
@@ -197,6 +197,7 @@ describe("getConfig", () => {
       uploadGitMetadata: true,
       failOnError: false,
       skipCloudformationOutputs: false,
+      mergeStepFunctionAndLambdaTraces: false,
```
Comment (nine5two7) [inline src/env.spec.ts:200]: Should be default true or false. I prefer true for but we can discuss during the stand-up meeting.

File: src/forwarder.spec.ts [hunk 2]
```diff
@@ -1412,3 +1421,116 @@ describe("addStepFunctionLogGroupSubscription", () => {
     `);
   });
 });
+
```
[the remark this answers is not in the document]
Comment (kimi-p) [reply src/forwarder.spec.ts:1424]: Thanks! I'll use it when next time when the input is smaller.

File: src/env.spec.ts (hunk 1 above)
Comment (kimi-p) [reply src/env.spec.ts:200]: Good suggestion! Lemme verified with Sumedha.

**Review by nine5two7 — APPROVED** [review]
LGTM

File: README.md [hunk 3]
```diff
@@ -25,44 +25,45 @@ Each version of the plugin is published with a [specific set of versions of the
...
+| `apmFlushDeadline`            | Used to determine when to submit spans before a timeout occurs, in milliseconds. When the remaining time in an AWS Lambda invocation is less than the value set, the tracer attempts to submit the current active spans and all finished spans. Supported for NodeJS and Python. Defaults to `100` milliseconds.                                                                                                            |
+| `mergeStepFunctionAndLambdaTraces` | When set to `false`, Lambda trace will not merge with the calling Step Functions trace. Default to `true`.                                                                                                                                                                                                                                                                                                                  |
```
Comment (buraizu) [inline README.md:66]:
```suggestion
| `mergeStepFunctionAndLambdaTraces` | When set to `false`, Lambda traces do not merge with the calling Step Functions trace. Defaults to `true`.                                                                                                                                                                                                                                                                                                                  |
```

**Review by buraizu — APPROVED** [review]
Thanks, approving with some minor updates for consistency and per the [style guide](https://github.com/DataDog/documentation/blob/master/CONTRIBUTING.md#tense).

File: README.md (hunk 3 above)
Comment (kimi-p) [reply README.md]: Updated. Thanks for the comment!

**Comment by kimi-p** [comment]
Verified again that trace and definition are both as expected after flag name change (to singular step function)
image

image
(2nd lambda is using legacy api)

**Pull request merged**